flink-table/flink-sql-parser/src/main/java/org/apache/calcite/sql/ExtendedSqlRowTypeNameSpec.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.calcite.sql;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.util.Litmus;
+import org.apache.calcite.util.Pair;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A sql type name specification of ROW type.
+ *
+ * <p>The difference with {@link SqlRowTypeNameSpec}:
+ * <ul>
+ *   <li>Support comment syntax for every field</li>
+ *   <li>Field data type default is nullable</li>
+ *   <li>Support ROW type with empty fields, e.g. ROW()</li>
+ * </ul>
+ */
+public class ExtendedSqlRowTypeNameSpec extends SqlTypeNameSpec {
```

Comment:
is it better to inherit from `SqlRowTypeNameSpec`?

Reply:
`SqlRowTypeNameSpec` has assertion that the row fields be must more than 1, so i can not inherit from it, because we allow empty row: `ROW()`.